pkg/bpf/map.go

```diff
@@ -159,68 +155,34 @@ func (d DesiredAction) String() string {
 	}
 }
 
-// mapTypeToFeatureString maps a MapType into a string defined by run_probes.sh
-func mapTypeToFeatureString(mt MapType) string {
-	var featureString string
-	switch mt {
-	case MapTypeLPMTrie:
-		featureString = fmt.Sprintf("#define HAVE_LPM_MAP_TYPE")
-	case MapTypeLRUHash:
-		featureString = fmt.Sprintf("#define HAVE_LRU_MAP_TYPE")
-	default:
-		break
-	}
-	return featureString
-}
-
-// ReadFeatureProbes reads the bpf_features.h file at the specified path (as
-// generated by bpf/run_probes.sh), and stores the results of the kernel
-// feature probing.
-func ReadFeatureProbes(filename string) {
-	f, err := os.Open(filename)
-	if err != nil {
-		// Should not happen; the caller ensured that the file exists
-		log.WithFields(logrus.Fields{
-			logfields.Path: filename,
-		}).WithError(err).Fatal("Failed to read feature probes")
-	}
-	defer f.Close()
-	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
-		for mapType := MapTypeHash; mapType < MapTypeMaximum; mapType++ {
-			featureString := mapTypeToFeatureString(mapType)
-			if featureString != "" &&
-				bytes.Equal(scanner.Bytes(), []byte(featureString)) {
-				log.Debugf("Detected support for map type %s", mapType.String())
-				supportedMapTypes[mapType] = true
-			}
-		}
-	}
-
-	for mapType := MapTypeHash; mapType < MapTypeMaximum; mapType++ {
-		if mapTypeToFeatureString(mapType) == "" {
-			log.Debugf("Skipping support detection for map type %s", mapType.String())
-		} else if _, probed := supportedMapTypes[mapType]; !probed {
-			log.Debugf("Detected no support for map type %s", mapType.String())
-			supportedMapTypes[mapType] = false
-		}
-	}
-}
-
 // GetMapType determines whether the specified map type is supported by the
 // kernel (as determined by ReadFeatureProbes()), and if the map type is not
 // supported, returns a more primitive map type that may be used to implement
 // the map on older implementations. Otherwise, returns the specified map type.
-func GetMapType(t MapType) MapType {
+func GetMapType(t MapType) (MapType, error) {
+	var err error
+	getMapTypes := func() {
+		var probeManager *probes.ProbeManager
+		probeManager, err = probes.NewProbeManager()
```

Comment:
From the above two comments, I'm wondering if it'd actually be more reliable to just do the `sync.Once()` logic inside `NewProbeManager()` to store the results of the initial probing and reliably either fail on startup or succeed always thereafter. I'm nervous that some weird condition might only cause the second execution of probing to fail and that leading to other weird behaviours, better to just do it once at the start.

This function would then just always fetch the probe manager and grab the `supportedMapTypes`, it wouldn't need to cache it as a package-level variable.

Reply:
This is a good idea. I tried it and the code is much cleaner. Please take a look again.